### PR TITLE
[onert] Use original map in ConstantInsertionPass

### PR DESCRIPTION
**Pull request by hseok-oh**
This commit add original operand map to use original operand if backend layout is same. 
It will prevent removing original operand and reduce constant operand creation.

ONE-DCO-1.0-Signed-off-by: Hyeongseok Oh <hseok82.oh@samsung.com>

### CODE REVIEW

**Comment by ragmani** [comment]
(Out of curiosity)
As far as I know, `ConstantInsertionPass` has had one role that inserts operands only when circle model has shared constant.
Now, are you going to allow `ConstInsertionPass` to insert all required operand too.

File: runtime/onert/core/src/compiler/pass/ConstantInsertionPass.cc [hunk 1]
```diff
@@ -40,9 +40,25 @@ void ConstantInsertionPass::callback(const ir::OperationIndex &node_index, ir::I
...
+    if (_keep_operands_map.find(input) == _keep_operands_map.end())
+    {
+      _keep_operands_map.emplace(input, factor);
+      continue;
+    }
+
+    // Same backend and layout with original operand. Keep using original operand
+    if (_keep_operands_map.at(input) == factor)
+      continue;
+
```
Comment (glistening) [inline runtime/onert/core/src/compiler/pass/ConstantInsertionPass.cc:59]: (optional)

Here is my trial to use `insert`'s return value — [`std::pair<iterator, bool>`](http://en.cppreference.com/w/cpp/utility/pair).

It is just for recording. I am not sure the code is better.

### 1st

`.second` seems not readable.

```suggestion
    if (_keep_operands_map.insert({input, factor}).second || _keep_operands_map.at(input) == factor)
      continue;
```

### 2nd

Thus, I've used c++17 features:
- if statements with initializer
- structured binding

```suggestion
    if (auto const [existing, ok] = _keep_operands_map.insert({input, factor}); ok || existing->second == factor)
      continue;
```

It requires an update in `CMakeLists.txt`.

```
set(CMAKE_CXX_STANDARD 17)
```

**Comment by glistening** [comment]
@hseok-oh What happens if `input` has 3 uses, and 2nd and 3rd have the same factor? Also, AFAIK, `OperationPass` iterate operation via `_graph.operations().iterate`. It seems not be guaranteed topological_sort or something. Is it okay to iterate by default order?

**Review by glistening — DISMISSED** [review]
LGTM

File: runtime/onert/core/src/compiler/pass/ConstantInsertionPass.cc [hunk 2]
```diff
@@ -40,9 +40,25 @@ void ConstantInsertionPass::callback(const ir::OperationIndex &node_index, ir::I
     auto &object = _graph.operands().at(input);
 
     const auto key = ReplaceKey{input, factor};
-    if (object.isConstant() && (object.getUses().size() >= 2 ||
-                                _replace_operands_map.find(key) != _replace_operands_map.end()))
+
+    // Skip if the operand is not constant or has only one use
+    if (!object.isConstant() || object.getUses().size() < 2)
+      continue;
```
Comment (hseok-oh) [inline runtime/onert/core/src/compiler/pass/ConstantInsertionPass.cc:46]: @ragmani 

>As far as I know, ConstantInsertionPass has had one role that inserts operands only when circle model has shared constant.
From now on, are you going to allow ConstInsertionPass to insert all required constant operands as well.

Constant insertion is skipped if constant is not shared (`object.getUses().size() < 2`)

**Comment by hseok-oh** [comment]
I've updated comment

File: runtime/onert/core/src/compiler/pass/ConstantInsertionPass.cc (hunk 2 above)
Comment (hseok-oh) [reply runtime/onert/core/src/compiler/pass/ConstantInsertionPass.cc]: > are you going to allow ConstInsertionPass to insert all required constant operands as well.

I didn't think about usage of this pass. I updated this to handle all operand possible case for easy understood of this pass's objective.

File: runtime/onert/core/src/compiler/pass/ConstantInsertionPass.cc [hunk 3]
```diff
@@ -40,9 +40,27 @@ void ConstantInsertionPass::callback(const ir::OperationIndex &node_index, ir::I
     auto &object = _graph.operands().at(input);
 
     const auto key = ReplaceKey{input, factor};
```
Comment (glistening) [inline runtime/onert/core/src/compiler/pass/ConstantInsertionPass.cc:42]: @hseok-oh 

I cannot find my previous comment. Maybe it is not written.

The declaration of `key` is used immediately after this line before this PR.

With this change, the use of `key` is a far below.

Please move line 42 to immediately before line 64.

Comment (hseok-oh) [reply runtime/onert/core/src/compiler/pass/ConstantInsertionPass.cc:42]: Right. I did not move this to reduce code change.

Comment (hseok-oh) [reply runtime/onert/core/src/compiler/pass/ConstantInsertionPass.cc:42]: I'll update

File: runtime/onert/core/src/compiler/pass/ConstantInsertionPass.cc (hunk 2 above)
Comment (ragmani) [reply runtime/onert/core/src/compiler/pass/ConstantInsertionPass.cc]: > I updated this to handle all operand possible case for easy understood of this pass's objective.

It looks good to me.

**Review by glistening — APPROVED** [review]
LGTM

**Review by ragmani — APPROVED** [review]
LGTM

**Review by jyoungyun — APPROVED** [review]
LGTM

**Pull request merged**